README.md
**CalliopEO**

***[README](README.md)*** | [Program Description](ProgramDescription.md) | [Testcases](testcases/testcases.md) | [Changelog](CHANGELOG.md)
# Readme

## Description
CalliopEO is a Python script to facilitate interaction between the [Astro Pi microcomputer](https://astro-pi.org/) and the [CalliopEO Mini microcontroller board](https://calliope.cc/en/calliopeo). If executed, the script detects, if a CalliopEO Mini is attached to a USB board of the Astro Pi and determins the serial port to communicate with the CalliopEO.

The directory, in which the `CalliopEO.py` script is located is called the main directory. Place any program(s) to be executed by the CalliopEO as zipped HEX file(s) in this main directory. If executed the script will search for all zip archives, unpack the CalliopEO program (HEX file) from the archive and flash the CalliopEO with this program. After flashing, the CalliopEO will reboot automatically and execute the program.

In the main directory a sub-folder named `run_YYYYMMDD-HHMMSS` will be created. The HEX files flashed and executed on the CalliopEO will be copied to this folder along with any data sent back by the program (files will end with `.data`). The initial zip archive in the main folder is renamed (additional suffix `.done`) to exclude this file from being processed again.

The `CalliopEO.py` script can collect data sent by the program on the CalliopEO via the USB serial port. Therefore, prepare the CalliopEO program to wait for the string `@START@`. Then, the CalliopEO program should respond by sending `@START@` back to the `CalliopEO.py` script and only after this sending the data. After sending the data the CalliopEO program should send the
message `@END@`.

Each line in the `.data` files will have a leading time stamp of the format `YYYY/MM/DD-HH/MM/SS.ssssss`. A boilerplate for the CalliopEO programs is described below.

## Execute
### Important considerations when using on Astro Pi onboard the ISS
Due to the movement of the ISS around the Earth a loss of signal (LOS) between a Ground Control Center and the Astro Pi can occur. This can result in a loss of the running SSH session running `CalliopEO.py`. To ease the process of gaining back access, it is highly recommended to use a terminal multiplexer. On Astro Pi OS the screen multiplexer [`screen`](https://www.gnu.org/software/screen/) is available.

To establish a terminal session via `screen` that can be resumed after a LOS follow the following steps:
```
# Establish a SSH connection to Astro Pi with the hostname or IP address `<astropi>` using a SSH-enabled user account on Astro Pi `<ssh-user>`. Replace both `<astropi>` and `<ssh-user>` with the correct values:
ssh <ssh-user>@<astropi>
# Change to the user calliope and change to it's home directory
su - calliope
# List all currently running screen sessions:
screen -ls
```
With multiple sessions the ouput could be similar to:
```
There are screens on:
    10474.calliope1 (Detached)
    10427.testing   (Detached)
    10321.other_session     (Detached)
3 Sockets in /var/run/screen/S-thiht.
```
To connect to an existing but detached session you can use the Process ID (PID, first collumn) or the session name (second column). For example:
```
screen -x calliope1
```
In case a new screen session has to be created, execute the followinf command and select a propper session name `<name>`:
```
screen -S <name>
```
At any time a running session can be detached using the keyboard sequence `CTRL+a, d`. Resume a session detached either manually or by a LOS using `screen -x <name>` (see above).

### Command Syntax of CalliopEO.py
The syntax of the script `CalliopEO.py` is:
```
$ ./CalliopEO.py [--max-data-size=bytes] [--max-script-execution-time=seconds] [--fake-timestamp]
```
* `--max-data-size` is the maximum number of characters to be read from the CalliopEO.
* With `--max-script-execution-time` you can specify a maximum time to accept input from the CalliopEO program before terminating the connection.
* `--fake-timestamp` adds to the beginning of each line a constant time stamp `2000/01/01-00:00:00.000000` instead of the current time stamp. This feature is primarily meant for testing purposes.

## Hardware Setup
The CalliopEO should be connected via USB to the Astro Pi.
```
     +--------------+
     |              |      USB      o---o---o
     |   Astro Pi   |===============| Call. |
     |              |               o---o---o
     +--------------+                
```

## Testing

![Selection of testcases](assets/select_testcases.png)

To perform software system tests a suite of testcases were established. To select and run individual testcases execute `./testing.sh`. The following dialog allows to select or deselect individual testcases, but the order of tests cannot be changed. To get an overview of available testcases see [Testcases.md](testcases/testcases.md).

Important: Before executing `./testing.sh` please remove all existing ZIP files
and `run_` folders from the main directory. `./testing.sh` will abort if those
files or diretories are present.

The test suite is primarily meant to support the development and maintenance of
the software.

Sample testresults can be found in [./testresults](testresults).

## HEX Boilerplate

For a sample implementation of the Boilerplate code for the CalliopEO please have a look at the [JS File](https://github.com/calliope-edu/CalliopEO_AstroPi/blob/06a4867b96a7cef7bf293340d21f18b37555aabc/testcases/testfiles/30sec-iss-sensors.js), the [Compiled HEX ](https://github.com/calliope-edu/CalliopEO_AstroPi/blob/06a4867b96a7cef7bf293340d21f18b37555aabc/testcases/testfiles/30sec-iss-sensors.hex) or edit it on [makecode.calliope.cc](https://makecode.calliope.cc/_KJT7WkEcwaDi). The .data file on the Astro Pi should look like this [this](https://github.com/calliope-edu/CalliopEO_AstroPi/blob/06a4867b96a7cef7bf293340d21f18b37555aabc/testcases/testfiles/30sec-iss-sensors.hex.data)

## Software Installation
### Operating system
The project is intended to be used with the ESA [Astro Pi IR](https://astro-pi.org/) onboard the International Space Station ISS. The Astro Pi SBCs are running a dedicated flavour of Raspberry Pi OS not available for the public. But this software can also be installed on the publicly available [Raspberry Pi OS](https://www.raspberrypi.org/software/) running on any Raspberry Pi, but the installation and operation is not covered by this document. 

### Installation using the setup.sh script
To set up the environment for the `CalliopEO.py` script a setup script `setup.sh` was established. It automatically performs all necesaary steps to provide a fully functional environment.

Before executing the setup script, connect the CalliopEO to the Astro Pi. Execute the setup script with the following command:
```
$ sudo setup.sh [username]
```
`username` is the name of the user to be created (enter without any brackets!). If this optional parameter is not given, the default `calliope` is used.

The setup script performs the following actions:
* Create a dedicated user (default: `calliope`). If the selected user already exists, `setup.sh` will exit.
* Add user to group `dialout` to be able to perform communication with the CalliopEO over the serial port
* Set password for the new user
* Determine UUIDs for CalliopEO's block devices and creating mount point in `/etc/fstab`. Later, this is used to mount the block devices for flashing the CalliopEO. The mount point is defined in `/etc/fstab` in order to allow unprivileged users like `calliope` to mount/unmount the block device.
* Create the main directory below the user's home directory `~/calliopEO` and copy all necessary files to this directory, in particular the Python script `CalliopEO.py`.
* Installing necessary Python modules from local wheel files. This is necessary due to security requirements set by ESA.

## De-Installation
The system can be cleaned up from the files necessary and the system settings done exclusively for CalliopEO using the procedure described as below. For this procedure it is assumed, that the username for the CalliopEO user is `calliope`. Replace `calliope` by the appropriate username if an alternative username was selected during the setup process (see above).

1. Login to the Astro Pi using a SSH-enabled user different from `calliope`.
2. Remove the user `calliope` with the command
```
$ sudo userdel -r calliope
```
3. Verify that the above command successfully removed the home directory `/home/calliope` with the command `ls /home/`. If the home directory still exists, remove it using the command `$ sudo rm -rf /home/calliope`.
4. Create a backup of the file `/etc/fstab` using the command
```
$ sudo cp /etc/fstab /etc/fstab.backup
```
5. Open the file `/etc/fstab` in a text editor. Remove (or uncomment) mount point definitions in `/etc/fstab` for CalliopEO. The line(s) with the moint point definitions should look like the following:
```
# Mount points for CalliopEO
/dev/disk/by-uuid/0123-4567 /home/calliope/mnt/mini vfat noauto,users 0 0
```
6. Restart the operating system. The de-installation is complete.


CHANGELOG.md
**CalliopEO**

[README](README.md) | [Program Description](ProgramDescription.md) | [Testcases](testcases/testcases.md) | ***[Changelog](CHANGELOG.md)***
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Tracking changes in master

*no changes yet*

## 1.3.0

- Added `infinite-iss-sensors-2s` `.hex` and `.data`
- Described more elegant way to change to user calliope
- Removed configuration for serial port, see issue #86
- provide result of iss sensors as csv
- increase MAX_LINE_LENGTH to 1000
- Switch display on and of with A button in iss-sensors.hex
- use compass in iss-sensors.hex without calibration
- optional calibration of compas by pressing A+B in iss-sensors.hex
- update .data files
- update testresults

## 1.2.0

**New**
- Checking now for necessary programs `lsblk`, `md5sum`, `sum` and `pip3` at installation
- add check for programs `lsblk`, `md5sum`, `sum` and `pip3` to `testcase 01`
- Exit setup.sh if login shell is provided on serial port (issue #86)

**Updates**
- Update testresults

**Fixes**
- replaced `Raspberry Pi` references by `Astro Pi`
- replaced `CalliopEO` by `Calliope mini` in readme
- Removed wheel for argparse because it is already installed on AstroPi-IR (issue #87)
- Correct syntax for running `./CalliopEO.py`
- removed root user refference in Readme

## [1.1.1](https://github.com/calliope-edu/CalliopEO_AstroPi/releases/tag/v1.1.1)

**Fixes**
- Fix broken links A2 & A3 in ProgramDescription

## [1.1.0](https://github.com/calliope-edu/CalliopEO_AstroPi/releases/tag/v1.1.0)

**Updates**
- Closed issue #78: Testing.sh exits with notification if there are no test case files
- Closed issue #76: Inserted "5 second sleep phase" between test cases with unconnected and connected CalliopEO in testing.sh to avoid that the program accesses CalliopEO too early which can result in errors
- Closed issue #74: Reduced data rate for burst.hex to avoid having corrupted .data files resulting in incorrect MD5 checksums
- Closed issue #48: Handle "unstructured" data from CalliopEO without newlines
- Add test case for data without newline
- Update Test results
- Add error message output for `Max file size achieved` and `Max script time achieved`
- Add MIT License
- Do not add FLASH device in `/etc/fstab` in setup process anymore
- Updates in `Readme` and `ProgramDescription`
- Update to initial Hex file `30s-iss-sensors.hex`
- Replaced md5sum by custom-made `comp()` function in test cases 07a and 07b
- Removed unneeded declaration of `tc_folder`
- Added Cancel button to GUI
- Max number of retries to flash can now be configured via `MAX_RETRY_FLASHING`
- Max serial line length read can now be configured via `MAX_LINE_LENGTH`
- Cleaned up function `waitSerialStart()`, see issue #56
- Properly encode SERIAL_START before sending to serial port
- Added notes on proper usage of `SERIAL_TIMEOUT` and `MAX_SERIAL_WAIT_REPLY`


## [1.0.1](https://github.com/Amerlander/svelte-typeahead-multiselect/releases/tag/v1.0.0) - 2021-02-25

**Updates**

- Link Boilerplate files in Readme
- Describe CLI Argument --fake-timestamp in Readme

## [1.0.0](https://github.com/calliope-edu/CalliopEO_AstroPi/releases/tag/v1.0.0) - 2021-07-24

- Initial release


CalliopEO.py
#!/usr/bin/env python3

import os
import shutil
from datetime import datetime, timedelta
import sys
import serial
import serial.tools.list_ports
import time
import blkinfo
import re
import argparse

blk = blkinfo.BlkDiskInfo()

TEMP_MOUNT_MINI = "~/mnt/mini"
TEMP_MOUNT_FLASH = "~/mnt/flash"

# Expand mount path if they contain tilde
TEMP_MOUNT_MINI = os.path.expanduser(TEMP_MOUNT_MINI)
TEMP_MOUNT_FLASH = os.path.expanduser(TEMP_MOUNT_FLASH)

MODEL_MINI_REGEXP = 'SEGGER[-_ ]{1}MSD[-_ ]{1}Volume'
MODEL_FLASH_REGEXP = 'SEGGER[-_ ]{1}MSD[-_ ]{1}FLASH'

# Compile regexp match pattern objects
MODEL_MINI_MPO = re.compile(MODEL_MINI_REGEXP, re.IGNORECASE)
MODEL_FLASH_MPO = re.compile(MODEL_FLASH_REGEXP, re.IGNORECASE)

MSG_MINI_NOT_FOUND = "mini not found"
MSG_FLASH_NOT_FOUND = "flash not found"
#MSG_MINI_NO_SERIAL = "no serial connection found"

CMD_MOUNT = "mount %s"
CMD_UNMOUNT = "umount %s"
CMD_SYNC = "sync %s"

#define the archive ending you want to search for
archive_ending = ".zip"

DEFAULT_ENCODING = "utf-8"
MINI_SERIAL_REGEXP = "(/dev/tty[\w]+)[ -]+Calliope mini[ -]+CDC"
MINI_SERIAL_MPO = re.compile(MINI_SERIAL_REGEXP, re.IGNORECASE)
SERIAL_START = "@START@"
SERIAL_END = "@END@"
# SERIAL_TIMEOUT and MAX_SERIAL_WAIT_REPLY are different! SERIAL_TIMEOUT is
# the timeout defined for operations performed on the pySerial object.
# MAX_SERIAL_WAIT_REPLY is the maximum time in seconds to wait for a propper
# reply on the serial port after sending the start identifier SERIAL_START.
# SERIAL_TIMEOUT << MAX_SERIAL_WAIT_REPLY
SERIAL_TIMEOUT = 1 # s
MAX_SERIAL_WAIT_REPLY = 10 # Max time in s to wait for answer
MAX_RETRY_FLASHING = 3 # Max. number to retry flashing if no serial data
MAX_CONNECTION_TIME = 11100 # s
MAX_DATA_SIZE = 20 * 1024 * 1024 # Bytes

MAX_LINE_LENGTH = 1000 # Crop lines that are longer

# Returns the port for the (first) Calliope mini or None
def getMiniSerial():
    # Retrieve a list of serial ports
    # The serial port for Calliope Mini is typically
    # identified via "/dev/ttyACM0 - Calliope Mini - CDC"
    all_ports = serial.tools.list_ports.comports()

    # Iterate through all serial ports and return the first port
    # found where a Calliope Mini is attached to
    mini_port = None # Initialize
    for p in all_ports:
        m = re.match(MINI_SERIAL_MPO, str(p))
        if m is not None:
            mini_port = m.group(1)
            break

    return mini_port

# Connect to serial port and return pySerial instance. If connection
# is not successful, returns None
def serialConnect(mini_port):
    # Create pySerial instance with all the necessary parameters but
    # with port = None. This way, the port is not opened immediately
    ser = serial.Serial(
        port = None,
        baudrate = 115200,
        bytesize = serial.EIGHTBITS,
        parity = serial.PARITY_NONE,
        stopbits = serial.STOPBITS_ONE,
        timeout = SERIAL_TIMEOUT)

    # try to connect
    try:
        ser.port = mini_port
        if not ser.is_open:
            ser.open()
        # Flush incoming buffer of serial port as a preventive measure
        # (see issue #74)
        ser.reset_input_buffer()
        mini_connected = True
    except Exception as Err:
        print("Error connecting serial port: %s" % Err)

    if mini_connected:
        return ser
    else:
        return None

# Listens to serial port, sends SERIAL_START and waits for response and
# data. Writes data with timestamps to file. Ends if timeout is reached,
# maximum size of data is written to file or SERIAL_END is received.
def readSerialData(ser, outFileName, fake_timestamp=False):

    # Initialize buffer for serial data
    serial_buffer = ''

    # Determine start/stop time for connection
    connStartTime = datetime.now()
    connEndTime = (connStartTime
            + timedelta(seconds=MAX_CONNECTION_TIME))
    connTimeoutNoResponse = (connStartTime
            + timedelta(seconds=MAX_SERIAL_WAIT_REPLY))

    # Write info to STDOUT
    tformat="%Y/%m/%d-%H:%M:%S"
    print("\r\n"
            + "Start @ "
            + connStartTime.strftime(tformat)
            + "; Will stop @ "
            + connEndTime.strftime(tformat)
            )

    # connTimeout can happen for two reasons:
    #   1. Still receivedStart == False after MAX_SERIAL_WAIT_REPLY
    #   2. receivedStart == True and serial connection established for
    #      longer than connEndTime
    connTimeout = False
    receivedStart = False
    # Variable dataSize is updated with the data in bytes written to outfile.
    # If dataSize exceeds MAX_DATA_SIZE no more data is written to file and
    # the function is exited.
    dataSize = 0

    # Main while loop
    while (connTimeout == False):

        # If receivedStart == False send SERIAL_START
        if (receivedStart == False):
            print("\r\n" + "Sending " + SERIAL_START)
            ser.write((SERIAL_START + '\r\n').encode(DEFAULT_ENCODING))
            time.sleep(1)

        # Append incoming data from serial port to serial_buffer
        if ser.in_waiting > 0:
            try:
                incoming = ser.read(ser.in_waiting).decode(DEFAULT_ENCODING)
            except:
                incoming = ''

            serial_buffer += incoming

            # Check if serial_buffer contains newline character. If so,
            # extract the first line for further processing. The function
            # .split('\n', 1) returns a list with one element if
            # serial_buffer contains no newline character and a list with 2
            # elements if there are newline characters.
            buffer_split = serial_buffer.split('\n', 1)

            # If serial_buffer contains no newline but the length exceeds
            # MAX_LINE_LENGTH, than extract MAX_LINE_LENGTH from
            # serial_buffer.
            if (len(buffer_split) == 1 and len(serial_buffer) >
                MAX_LINE_LENGTH):
                buffer_split = [
                    serial_buffer[:MAX_LINE_LENGTH],
                    serial_buffer[MAX_LINE_LENGTH:],
                    ]

            if (len(buffer_split) == 2):
                # This is the current line. The line does not end with
                # newline
                line = buffer_split[0].strip()

                # If receivedStart == False check if the current line
                # contains SERIAL_START and update receivedStart if
                # necessary
                if (receivedStart == False):
                    # If receivedStart == False and line contains
                    # SERIAL_START than set receivedStart to True and do
                    # nothing else with line
                    if (SERIAL_START in line):
                        print("\r\n" + "Received " + SERIAL_START)
                        receivedStart = True
                else:
                # receivedStart == True: Process line normally
                    print("*",end="",flush=True)

                    # If line contains SERIAL_END we exit the while loop
                    if (SERIAL_END in line):
                        break

                    # Add time stamp to beginning of line (Github issue #45).
                    # If --fake-timestamp is set, then set the time stamp to
                    # constant value 2000/01/01-00:00:00.000000
                    if (fake_timestamp == True):
                        ts = "2000/01/01-00:00:00.000000"
                    else:
                        ts = datetime.now().strftime("%Y/%m/%d-%H:%M:%S.%f")
                    line = ts + " " + line

                    # Output to file if updated dataSize plus length of lines
                    # plus 1 (for newline) do not exceed MAX_DATA_SIZE
                    if ((dataSize + len(line) + 1) <= MAX_DATA_SIZE):
                        write2File(outFileName, line+'\n')
                        dataSize += (len(line) + 1)
                    else:
                        # Exit while loop
                        print("\r\n" + "Max file size achieved")
                        break

                # Update serial_buffer
                serial_buffer = buffer_split[1]

        # Check time and update connTimeout if necessary
        now = datetime.now()
        # See if receivedStart == False and we are behind
        # connTimeoutNoResponse
        if (receivedStart == False and now > connTimeoutNoResponse):
            connTimeout = True
        # See if we are behind connEndTime
        if (receivedStart == True and now > connEndTime):
            print("\r\n" + "Max script time achieved")
            connTimeout = True

    return dataSize

def listFolders(path):
    temp_list = []
    for folder in os.listdir(path):
        if os.path.isdir(os.path.join(path, folder)):
            temp_list.append(os.path.join(path, folder))
    return temp_list

#unpacks given archives in a folder and returns the foldername
def unpackArchives(archive_list, destname=None):
    if(destname == None):
        archive_folder = os.path.join(
                os.getcwd(),
                datetime.now().strftime("run_%Y%m%d-%H%M%S")
                )
    else:
        archive_folder = os.path.join(os.getcwd(),destname)

    # Create empty archive_folder
    os.mkdir(archive_folder)

    for file in archive_list:
        try:
             shutil.unpack_archive(file, archive_folder)
             shutil.move(file, file + ".done")
             print("Unpacked " + file)
        except Exception as Err:
            shutil.move(file, file + ".failed")
            print("Error while unpacking " + file + ": %s" % Err)
            # Wait a second. Otherwise the folder run_YYYYMMDD-HHMMSS
            # for the next zip archive can have the time stamp which
            # would result in trying to create the same diretory which
            # in the end will cause the script to fail (Github issue #34).
            time.sleep(1)

    return archive_folder

#list files in this path filtered by ending and recurse if desired
def listFiles(path, ending=None, recurse=False):
    temp_list = []
    for file in os.listdir(path):
        if os.path.isfile(os.path.join(path,file)):
            if ending == None:
                temp_list.append(os.path.join(path,file))
            else:
                if file.endswith(ending):
                    temp_list.append(os.path.join(path,file))
        else:
            if recurse and os.path.isdir(os.path.join(path,file)):
                temp_list.extend(listFiles(os.path.join(path,file), ending))
    return sorted(temp_list)

def getMiniDisk():
    MINI_DEVICE = ""
    disks = blk.get_disks()
    for disk in disks:
        if MODEL_MINI_MPO.match(disk['model']) is not None:
            return True
    return False

def getFlashDisk():
    FLASH_DEVICE = ""
    disks = blk.get_disks()
    for disk in disks:
        if MODEL_FLASH_MPO.match(disk['model']) is not None:
            return True
    return False

#programm mini
def programmMini(hex):
    #mount mini disk
    os.system(CMD_MOUNT % (TEMP_MOUNT_MINI))
    #programm mini
    shutil.copy2(hex, TEMP_MOUNT_MINI)
    os.system(CMD_SYNC % TEMP_MOUNT_MINI)
    os.system(CMD_UNMOUNT % TEMP_MOUNT_MINI)

def write2File(outFileName, string):
    # Open file in append mode
    outfile = open(outFileName,"a")
    # Write strings, NOT lines! Care yourself for newlines!
    outfile.write(string)
    outfile.close()

###################################################

def main(args):
    print("-=# CalliopEO #=-")

    # If CLI paramater --max-data-size or --max-script-execution-time
    # are set, update the global variables
    if args.max_data_size > 0:
        global MAX_DATA_SIZE
        MAX_DATA_SIZE = args.max_data_size
    if args.max_script_execution_time > 0:
        global MAX_CONNECTION_TIME
        MAX_CONNECTION_TIME = args.max_script_execution_time

    #check mini disk
    if not getMiniDisk():
        print(MSG_MINI_NOT_FOUND)
        sys.exit(10)
    #check flash disk
    if not getFlashDisk():
        print(MSG_MINI_NOT_FOUND)
        sys.exit(11)
    #make mini mount dir
    if not os.path.exists(TEMP_MOUNT_MINI):
        os.mkdir(TEMP_MOUNT_MINI)
    #make flash mount dir
    if not os.path.exists(TEMP_MOUNT_FLASH):
        os.mkdir(TEMP_MOUNT_FLASH)

    #get archives
    archive_file_list = listFiles(os.getcwd(), ending=archive_ending)
    if len(archive_file_list) == 0:
        print("no archives found in this directory")
        print("make sure you have at least one archive in this directory")
        sys.exit(12)
    #unpack archives to a folder named with current date and time
    folder_name = unpackArchives(archive_file_list)
    #recursively search for .hex files
    hex_files = listFiles(folder_name, ending=".hex" ,recurse=True)
    print("\r\n")

    for hex in hex_files:
        if os.path.exists(hex + ".data"):
            print("skipping: " + hex)
            continue
        count_try_flashing  = 1 # Max: MAX_RETRY_FLASHING
        while True:
            print("programming: " + hex)
            programmMini(hex)
            print("done")

            # After programming the Calliope Mini reboots. The serial ports
            # shows up quite soon. But it takes some time (~ 15 seconds) before
            # a read/write operation is possible, even if the Calliope Mini
            # accepted a connect (ser.open()).
            print("open serial port")
            ser = None
            mini_connected = False
            no_tries = 0
            while (not mini_connected and no_tries < 20):
                mini_port = getMiniSerial()
                if mini_port is not None:
                    print("Calliope Mini found on", mini_port, " ", end="", flush=True)
                    # Wait to give the Calliope Mini some time before
                    # connect. When conecting too early, the first read/write
                    # access will result in an I/O error.
                    for w in range(10):
                        print(".", end="", flush=True)
                        time.sleep(1)
                    ser = serialConnect(mini_port)
                    mini_connected = True
                else:
                    # Retry to connect after one second
                    no_tries += 1
                    time.sleep(1)

            print("\r\ndone")

            # If ser is still None at this point, connection to Calliope cannot
            # be established after flashing. Something serious might have
            # happend. Exit the script.
            if ser is None:
                print("\r\nCannot establish serial conection to Calliope Mini. Exiting.")
                sys.exit(13)

            print("reading data")
            outFileName = hex + '.data'
            dataSize = readSerialData(ser, outFileName, args.fake_timestamp)

            if dataSize == 0:
                if count_try_flashing >= MAX_RETRY_FLASHING:
                    #give up
                    break
                else:
                    #repeat programming
                    count_try_flashing += 1
                    print(
                            "\r\nSomething went wrong. Retrying flashing ("
                            + str(count_try_flashing)
                            + "/"
                            + str(MAX_RETRY_FLASHING)
                            + ")"
                            )
                    continue
            else:
                print("\r\ndone")
                print("############################################################################################################\r\n")
                break

# Exit status
# 0     normally
# 10    could not find mini disk (Calliope mini not attached)
# 11    could not find flash disk (Calliope mini not attached)
# 12    no archives in diretory
# 13    could not establish serial connection
if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument(
            '--max-data-size',
            default = 0,
            dest = 'max_data_size',
            type=int,
            )
    parser.add_argument(
            '--max-script-execution-time',
            default = 0,
            dest = 'max_script_execution_time',
            type=int,
            )
    parser.add_argument(
            '--fake-timestamp',
            action = 'store_true',
            dest = 'fake_timestamp',
            help = """
            Set timestamp in output to constant value
            2000/01/01-00:00:00.000000
            """,
            )
    args = parser.parse_args()
    main(args)
